binary operator raise exception

mdionisio opened the pull request:
it can be that node.left.name or node.right.name not exist

Traceback (most recent call last):
  File "/home/dionisio/.local/bin/autopxd", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/__init__.py", line 286, in cli
    outfile.write(translate(infile.read(), infile.name, extra_cpp_args, debug=debug, regex=regex))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/__init__.py", line 207, in translate
    p.visit(
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 266, in visit_Typedef
    decls = self.collect(node)
            ^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 284, in collect
    self.generic_visit(node)
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 200, in visit_TypeDecl
    decls = self.collect(node)
            ^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 284, in collect
    self.generic_visit(node)
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 158, in visit_Enum
    value_as_str, maybe_value_as_int = parse_enum_value(item.value)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 73, in parse_enum_value
    value_as_str = f"{node.left.name} {node.op} {node.right.name}"
                      ^^^^^^^^^^^^^^
AttributeError: 'Constant' object has no attribute 'name'

mdionisio commented:
I think that this pull request solve https://github.com/elijahr/python-autopxd2/issues/50

elijahr commented:
Thank you for the contribution! Can you add a test illustrating the issue, or at least provide some sample code to reproduce it?

Commit:
binary operator raise exception
it can be that node.left.name or node.right.name not exist

Traceback (most recent call last):
  File "/home/dionisio/.local/bin/autopxd", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/__init__.py", line 286, in cli
    outfile.write(translate(infile.read(), infile.name, extra_cpp_args, debug=debug, regex=regex))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/__init__.py", line 207, in translate
    p.visit(
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 266, in visit_Typedef
    decls = self.collect(node)
            ^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 284, in collect
    self.generic_visit(node)
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 200, in visit_TypeDecl
    decls = self.collect(node)
            ^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 284, in collect
    self.generic_visit(node)
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 93, in visit
    rv = super().visit(node)
         ^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 158, in visit_Enum
    value_as_str, maybe_value_as_int = parse_enum_value(item.value)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dionisio/.local/lib/python3.12/site-packages/autopxd/writer.py", line 73, in parse_enum_value
    value_as_str = f"{node.left.name} {node.op} {node.right.name}"
                      ^^^^^^^^^^^^^^
AttributeError: 'Constant' object has no attribute 'name'

mdionisio commented:
I update the test and also fix my patch.
